ngraph/core/src/op/topk.cpp

```diff
@@ -242,7 +242,12 @@ void op::v1::TopK::validate_and_infer_types()
             auto max_k = maximum_value(input_value(1));
             if (max_k.first)
             {
-                output_shape[m_normalized_axis] &= Dimension(0, max_k.second);
+                const auto in_min = output_shape[m_normalized_axis].get_min_length();
+                const auto in_max = output_shape[m_normalized_axis].get_max_length();
+                const auto lower = std::min<Dimension::value_type>(in_min, max_k.second);
+                const auto upper =
+                    in_max < 0 ? -1 : std::max<Dimension::value_type>(in_max, max_k.second);
```

Comment:
I don't like the idea to check that the value is equal to magic number `-1`. May be we should check that `output_shape[m_normalized_axis]` is dynamic instead?

@ilyachur , @GlebKazantaev , @tomdol , @mitruska, what do you think?

Reply:
In this case -1 is something that we get from here https://github.com/openvinotoolkit/openvino/blob/master/ngraph/core/src/dimension.cpp#L140 and this negative value is then used internally by this constructor https://github.com/openvinotoolkit/openvino/blob/master/ngraph/core/src/dimension.cpp#L50 to create an internal upper bound of a new Dimension object. I'm not sure if checking `output_shape[m_normalized_axis].is_dynamic()` would work the same way in the context of those instructions above. Can anyone else comment?

Reply:
I was also a little bit confused that [Dimension::get_max_length()](https://github.com/openvinotoolkit/openvino/blob/9c509e5f416cce29dddd3bd1de292055df3765c5/ngraph/core/src/dimension.cpp#L140) will return `-1` for the upper bound limit `Interval::s_max`.
It is possible to use `dimension.get_interval().get_max_val()` to get `Interval::s_max` value instead of `-1`.

Reply:
Can we use dynamic instead of -1?

Reply:
@ilyachur If you refer to `upper` variable, it can't, it has to be raw number. If you mean to set `output_shape[m_normalized_axis]` to fully dynamic then it'd be rolling back https://github.com/NervanaSystems/ngraph/pull/4512

Reply:
Yes I mean to use dynamic for `upper` value. Ca we use the next code?
```
const auto upper =
                    in_max < 0 ? Dimension::dynamic().get_max_length() : std::max<Dimension::value_type>(in_max, max_k.second);
```

Ot is it not correct?

Reply:
I've checked both solutions, as proposed by @mitruska and @ilyachur, and both work properly. 
I'd apply the later one as it relies on `Dimension` class only, while the former touches also `Interval` class.